File: test/server/Terminal.test.js
const {PassThrough} = require('stream')

// Terminal's module scope calls Odac.core('Log'); jest.setup.js installs the global mock.
const Terminal = require('../../server/src/Container/Terminal')

/**
 * Fake dockerode container.
 *
 * `exec()` serves two very different callers: the TTY session (Tty:true) and the throwaway
 * reap execs (Tty absent). Reap scripts are captured so tests can assert on the signal sent.
 */
function createContainerMock({runningSequence = [false]} = {}) {
  const sessionStream = new PassThrough()
  sessionStream.destroy = jest.fn(sessionStream.destroy.bind(sessionStream))

  const inspectResults = [...runningSequence]
  const sessionExec = {
    start: jest.fn(async opts => {
      sessionExec.startOptions = opts
      return sessionStream
    }),
    resize: jest.fn().mockResolvedValue(undefined),
    inspect: jest.fn(async () => {
      const running = inspectResults.length > 1 ? inspectResults.shift() : inspectResults[0]
      return {Running: running, ExitCode: running ? null : 0}
    })
  }

  const execOptions = []
  const reapScripts = []

  const container = {
    exec: jest.fn(async options => {
      execOptions.push(options)

      if (options.Tty) return sessionExec

      reapScripts.push(options.Cmd[2])
      return {
        start: async () => {
          const stream = new PassThrough()
          // End synchronously: Jest's fake timers also fake process.nextTick, so deferring
          // this would stall the reap's `await stream.on('end')` in the timeout tests.
          // 'end' still fires once #runInContainer attaches its drain listener.
          stream.end()
          return stream
        }
      }
    })
  }

  const docker = {getContainer: jest.fn(() => container)}

  return {docker, container, sessionExec, sessionStream, execOptions, reapScripts}
}

const sessionOptions = mock => mock.execOptions.find(o => o.Tty)

/**
 * Opens a session with the idle/lifetime timers off. Tests that leave a session open would
 * otherwise strand a 15-minute timer and keep the Jest worker alive. The timeout suite below
 * sets its own durations.
 */
const openTerminal = (mock, options = {}) => new Terminal(mock.docker, 'web', {idleTimeout: 0, maxLifetime: 0, ...options}).open()

describe('Terminal.open()', () => {
  test('creates a TTY exec sized up front and tagged with a session env var', async () => {
    const mock = createContainerMock()
    await openTerminal(mock, {cols: 120, rows: 40})

    const options = sessionOptions(mock)
    expect(options.Tty).toBe(true)
    expect(options.AttachStdin).toBe(true)
    // [h, w] — sizing at create avoids a first paint at 0x0.
    expect(options.ConsoleSize).toEqual([40, 120])
    expect(options.Env).toEqual([expect.stringMatching(/^ODAC_TTY_SESSION=[a-f0-9]{32}$/)])
  })

  test('repeats Tty on start, or the daemon streams multiplexed frames', async () => {
    const mock = createContainerMock()
    await openTerminal(mock)

    expect(mock.sessionExec.startOptions).toEqual({hijack: true, stdin: true, Tty: true})
  })

  test('defaults to a bash-or-sh probe and honours an explicit command', async () => {
    const mock = createContainerMock()
    await openTerminal(mock)
    expect(sessionOptions(mock).Cmd).toEqual(Terminal.DEFAULT_SHELL)

    const custom = createContainerMock()
    await openTerminal(custom, {command: ['/bin/zsh']})
    expect(sessionOptions(custom).Cmd).toEqual(['/bin/zsh'])
  })

  test('passes user, workdir and extra env through', async () => {
    const mock = createContainerMock()
    await openTerminal(mock, {user: 'app', workdir: '/srv', env: ['FOO=bar']})

    const options = sessionOptions(mock)
    expect(options.User).toBe('app')
    expect(options.WorkingDir).toBe('/srv')
    expect(options.Env).toEqual([expect.stringMatching(/^ODAC_TTY_SESSION=/), 'FOO=bar'])
  })

  test('refuses to open twice', async () => {
    const mock = createContainerMock()
    const terminal = await openTerminal(mock)

    await expect(terminal.open()).rejects.toThrow('already opened')
  })
})

describe('Terminal I/O', () => {
  test('forwards output to onData', async () => {
    const mock = createContainerMock()
    const chunks = []
    await openTerminal(mock, {onData: c => chunks.push(c.toString())})

    mock.sessionStream.write('hello')
    await new Promise(r => setImmediate(r))

    expect(chunks).toEqual(['hello'])
  })

  test('forwards input to the pty', async () => {
    const mock = createContainerMock()
    const terminal = await openTerminal(mock)

    const written = []
    mock.sessionStream.on('data', c => written.push(c.toString()))

    expect(terminal.write('ls\n')).toBe(true)
    await new Promise(r => setImmediate(r))

    expect(written).toEqual(['ls\n'])
  })

  test('resize maps cols/rows onto Docker w/h', async () => {
    const mock = createContainerMock()
    const terminal = await openTerminal(mock)

    await expect(terminal.resize(100, 30)).resolves.toBe(true)
    expect(mock.sessionExec.resize).toHaveBeenCalledWith({h: 30, w: 100})
  })

  test('resize clamps nonsense dimensions instead of passing them to Docker', async () => {
    const mock = createContainerMock()
    const terminal = await openTerminal(mock, {cols: 80, rows: 24})

    await terminal.resize(99999, 0)

    // width capped, height fell back to the previous value rather than 0
    expect(mock.sessionExec.resize).toHaveBeenCalledWith({h: 24, w: 1000})
  })

  test('resize survives a shell that exited mid-call', async () => {
    const mock = createContainerMock()
    const terminal = await openTerminal(mock)
    mock.sessionExec.resize.mockRejectedValueOnce(new Error('no such exec'))

    await expect(terminal.resize(90, 20)).resolves.toBe(false)
  })

  test('write and resize are no-ops once closed', async () => {
    const mock = createContainerMock()
    const terminal = await openTerminal(mock)
    await terminal.close()

    expect(terminal.write('x')).toBe(false)
    await expect(terminal.resize(90, 20)).resolves.toBe(false)
    expect(mock.sessionExec.resize).not.toHaveBeenCalled()
  })
})

describe('Terminal.close() reaping', () => {
  test('destroys the stream and hangs up the session tree', async () => {
    const mock = createContainerMock({runningSequence: [false]})
    const terminal = await openTerminal(mock)

    await terminal.close()

    expect(mock.sessionStream.destroy).toHaveBeenCalled()
    expect(mock.reapScripts).toHaveLength(1)
    expect(mock.reapScripts[0]).toContain('kill -HUP')
    expect(terminal.closed).toBe(true)
  })

  test('escalates to SIGKILL when the shell ignores the hangup', async () => {
    // Running: true after HUP, still true after the grace wait, false once KILLed.
    const mock = createContainerMock({runningSequence: [true, true, false, false]})
    const terminal = await openTerminal(mock)

    await terminal.close()

    expect(mock.reapScripts).toHaveLength(2)
    expect(mock.reapScripts[0]).toContain('kill -HUP')
    expect(mock.reapScripts[1]).toContain('kill -KILL')
  }, 10000)

  test('reaps as the session user, so a non-root shell can be read and signalled', async () => {
    // The reap scans /proc/<pid>/environ; reading a process owned by a different uid needs
    // CAP_SYS_PTRACE, which a stock container's exec lacks. Running the reaper as the shell's
    // own user keeps both the environ read and the kill on the same-uid path.
    const mock = createContainerMock({runningSequence: [false]})
    const terminal = await openTerminal(mock, {user: 'app'})

    await terminal.close()

    const reapOptions = mock.execOptions.find(o => !o.Tty)
    expect(reapOptions.User).toBe('app')
  })

  test('reap script targets only this session and cannot match itself', async () => {
    const mock = createContainerMock()
    await openTerminal(mock)

    const tag = sessionOptions(mock).Env[0].split('=')[1]
    await new Promise(r => setImmediate(r))
    mock.sessionStream.end() // triggers close('exited')
    await new Promise(r => setTimeout(r, 20))

    const script = mock.reapScripts[0]
    expect(script).toContain(`ODAC_TTY_SESSION=${tag}`)
    // Single line: `done \n | sort` is a shell syntax error.
    expect(script).not.toContain('\n')
    // The reaping exec is untagged, so the sweep can never signal itself.
    const reapExec = mock.execOptions.find(o => !o.Tty)
    expect(reapExec.Env).toBeUndefined()
  })

  test('reap signals children before the shell, so no zombies are left behind', async () => {
    const mock = createContainerMock({runningSequence: [false]})
    const terminal = await openTerminal(mock)

    await terminal.close()

    const script = mock.reapScripts[0]
    // Highest pid first; the shell is the oldest and lowest, and gets hung up last so it
    // can wait() on the children it just outlived. Killing it first strands them on pid 1.
    expect(script).toContain('sort -rn')
    expect(script.indexOf('kill -HUP "$p"')).toBeLessThan(script.indexOf('kill -HUP "$shell"'))
    expect(script).toMatch(/kill -HUP "\$p".*sleep 1.*kill -HUP "\$shell"/)
  })

  test('the session tag is generated internally, never taken from options', async () => {
    const mock = createContainerMock()
    // A caller-supplied id would land inside a shell command run in the container.
    await openTerminal(mock, {id: '"; rm -rf /; #', tag: 'evil'})

    expect(sessionOptions(mock).Env[0]).toMatch(/^ODAC_TTY_SESSION=[a-f0-9]{32}$/)
  })

  test('fires onExit exactly once with the exit code, even on repeated close', async () => {
    const mock = createContainerMock({runningSequence: [false]})
    const onExit = jest.fn()
    const terminal = await openTerminal(mock, {onExit})

    await terminal.close()
    await terminal.close()

    expect(onExit).toHaveBeenCalledTimes(1)
    expect(onExit).toHaveBeenCalledWith({reason: 'closed', exitCode: 0})
  })

  test('a shell that exits on its own closes the session as "exited"', async () => {
    const mock = createContainerMock({runningSequence: [false]})
    const onExit = jest.fn()
    const terminal = await openTerminal(mock, {onExit})

    mock.sessionStream.end()
    await new Promise(r => setTimeout(r, 20))

    expect(onExit).toHaveBeenCalledWith({reason: 'exited', exitCode: 0})
    expect(terminal.closed).toBe(true)
  })

  test('survives a container that vanished before the reap', async () => {
    const mock = createContainerMock({runningSequence: [false]})
    const onExit = jest.fn()
    const terminal = await openTerminal(mock, {onExit})

    mock.container.exec.mockRejectedValue(new Error('No such container'))

    await expect(terminal.close()).resolves.toBeUndefined()
    expect(onExit).toHaveBeenCalledTimes(1)
    expect(terminal.closed).toBe(true)
  })
})

describe('Terminal timeouts', () => {
  afterEach(() => jest.useRealTimers())

  test('closes an idle session and input pushes the deadline back', async () => {
    jest.useFakeTimers()
    const mock = createContainerMock({runningSequence: [false]})
    const onExit = jest.fn()
    const terminal = await openTerminal(mock, {onExit, idleTimeout: 1000, maxLifetime: 0})

    await jest.advanceTimersByTimeAsync(900)
    terminal.write('x') // resets the idle timer
    await jest.advanceTimersByTimeAsync(900)
    expect(onExit).not.toHaveBeenCalled()

    await jest.advanceTimersByTimeAsync(200)
    expect(onExit).toHaveBeenCalledWith(expect.objectContaining({reason: 'idle'}))
  })

  test('output alone does not keep an abandoned session alive', async () => {
    jest.useFakeTimers()
    const mock = createContainerMock({runningSequence: [false]})
    const onExit = jest.fn()
    await openTerminal(mock, {onExit, idleTimeout: 1000, maxLifetime: 0})

    await jest.advanceTimersByTimeAsync(600)
    mock.sessionStream.write('chatty process output')
    await jest.advanceTimersByTimeAsync(600)

    expect(onExit).toHaveBeenCalledWith(expect.objectContaining({reason: 'idle'}))
  })

  test('enforces a hard lifetime cap regardless of activity', async () => {
    jest.useFakeTimers()
    const mock = createContainerMock({runningSequence: [false]})
    const onExit = jest.fn()
    const terminal = await openTerminal(mock, {onExit, idleTimeout: 1000, maxLifetime: 2500})

    // Steady input keeps resetting the idle timer (last reset at 2400 → would fire at 3400),
    // so the cap at 2500 is the only thing that can end this session.
    for (let i = 0; i < 3; i++) {
      await jest.advanceTimersByTimeAsync(800)
      terminal.write('x')
    }
    expect(onExit).not.toHaveBeenCalled()

    await jest.advanceTimersByTimeAsync(200)
    expect(onExit).toHaveBeenCalledWith(expect.objectContaining({reason: 'lifetime'}))
  })

  test('a timeout of 0 disables the timer', async () => {
    jest.useFakeTimers()
    const mock = createContainerMock({runningSequence: [false]})
    const onExit = jest.fn()
    await openTerminal(mock, {onExit, idleTimeout: 0, maxLifetime: 0})

    await jest.advanceTimersByTimeAsync(60 * 60 * 1000)
    expect(onExit).not.toHaveBeenCalled()
  })
})

File: internal/docker/terminal_test.go
package docker

import (
	"bufio"
	"context"
	"net"
	"regexp"
	"strings"
	"sync"
	"testing"
	"time"

	"github.com/docker/docker/api/types"
	"github.com/docker/docker/api/types/container"

	"odac/internal/logx"
)

// termFakeAPI overrides the exec surface of fakeAPI with an interactive TTY
// model (jest's createContainerMock): the Tty exec is the session, untagged
// execs are reap sweeps whose scripts are captured.
type termFakeAPI struct {
	*fakeAPI

	tmu           sync.Mutex
	execOptions   []container.ExecOptions
	attachOptions []container.ExecAttachOptions
	reapScripts   []string
	resizes       []container.ResizeOptions

	// runningSeq answers ContainerExecInspect for the SESSION exec: shift
	// while more than one remains, then repeat the last (jest semantics).
	runningSeq []bool
	exitCode   int

	sessionServer net.Conn // write = pty output, read = pty input

	execCreateErr error
	resizeErr     error
}

func newTermFakeAPI() *termFakeAPI {
	return &termFakeAPI{fakeAPI: newFakeAPI(), runningSeq: []bool{false}}
}

func (f *termFakeAPI) ContainerExecCreate(_ context.Context, _ string, options container.ExecOptions) (container.ExecCreateResponse, error) {
	f.tmu.Lock()
	defer f.tmu.Unlock()
	if f.execCreateErr != nil {
		return container.ExecCreateResponse{}, f.execCreateErr
	}
	f.execOptions = append(f.execOptions, options)
	if options.Tty {
		return container.ExecCreateResponse{ID: "session"}, nil
	}
	f.reapScripts = append(f.reapScripts, options.Cmd[2])
	return container.ExecCreateResponse{ID: "reap"}, nil
}

func (f *termFakeAPI) ContainerExecAttach(_ context.Context, execID string, config container.ExecAttachOptions) (types.HijackedResponse, error) {
	f.tmu.Lock()
	defer f.tmu.Unlock()
	f.attachOptions = append(f.attachOptions, config)
	if execID == "session" {
		server, client := net.Pipe()
		f.sessionServer = server
		return types.HijackedResponse{Conn: client, Reader: bufio.NewReader(client)}, nil
	}
	// Reap exec: drain hits EOF immediately.
	server, client := net.Pipe()
	server.Close()
	return types.HijackedResponse{Conn: client, Reader: bufio.NewReader(client)}, nil
}

func (f *termFakeAPI) ContainerExecInspect(_ context.Context, execID string) (container.ExecInspect, error) {
	f.tmu.Lock()
	defer f.tmu.Unlock()
	running := f.runningSeq[0]
	if len(f.runningSeq) > 1 {
		f.runningSeq = f.runningSeq[1:]
	}
	inspect := container.ExecInspect{Running: running}
	if !running {
		inspect.ExitCode = f.exitCode
	}
	return inspect, nil
}

func (f *termFakeAPI) ContainerExecResize(_ context.Context, _ string, options container.ResizeOptions) error {
	f.tmu.Lock()
	defer f.tmu.Unlock()
	if f.resizeErr != nil {
		err := f.resizeErr
		f.resizeErr = nil
		return err
	}
	f.resizes = append(f.resizes, options)
	return nil
}

func (f *termFakeAPI) sessionOptions(t *testing.T) container.ExecOptions {
	t.Helper()
	f.tmu.Lock()
	defer f.tmu.Unlock()
	for _, o := range f.execOptions {
		if o.Tty {
			return o
		}
	}
	t.Fatal("no session exec created")
	return container.ExecOptions{}
}

func (f *termFakeAPI) reapOptions(t *testing.T) container.ExecOptions {
	t.Helper()
	f.tmu.Lock()
	defer f.tmu.Unlock()
	for _, o := range f.execOptions {
		if !o.Tty {
			return o
		}
	}
	t.Fatal("no reap exec created")
	return container.ExecOptions{}
}

func (f *termFakeAPI) scripts() []string {
	f.tmu.Lock()
	defer f.tmu.Unlock()
	return append([]string(nil), f.reapScripts...)
}

// exitRecorder collects onExit invocations.
type exitRecorder struct {
	mu    sync.Mutex
	calls []ExitInfo
}

func (r *exitRecorder) record(info ExitInfo) {
	r.mu.Lock()
	r.calls = append(r.calls, info)
	r.mu.Unlock()
}

func (r *exitRecorder) list() []ExitInfo {
	r.mu.Lock()
	defer r.mu.Unlock()
	return append([]ExitInfo(nil), r.calls...)
}

func durPtr(d time.Duration) *time.Duration { return &d }

// openTerminal opens a session with the timers off (a stranded 15-minute
// timer would outlive the test) and a fast reap grace.
func openTerminal(t *testing.T, f *termFakeAPI, options TerminalOptions) *Terminal {
	t.Helper()
	if options.IdleTimeout == nil {
		options.IdleTimeout = durPtr(0)
	}
	if options.MaxLifetime == nil {
		options.MaxLifetime = durPtr(0)
	}
	term := newTerminal(f, logx.New("Container", "Terminal"), "web", options)
	term.reapGrace = 5 * time.Millisecond
	if err := term.open(); err != nil {
		t.Fatal(err)
	}
	return term
}

func TestTerminalOpenCreatesTaggedTTYExec(t *testing.T) {
	f := newTermFakeAPI()
	term := openTerminal(t, f, TerminalOptions{Cols: 120, Rows: 40})
	defer term.Close("closed")

	options := f.sessionOptions(t)
	if !options.Tty || !options.AttachStdin || !options.AttachStdout || !options.AttachStderr {
		t.Errorf("attach flags: %+v", options)
	}
	// [h, w] — sizing at create avoids a first paint at 0x0.
	if options.ConsoleSize == nil || *options.ConsoleSize != [2]uint{40, 120} {
		t.Errorf("ConsoleSize = %v", options.ConsoleSize)
	}
	if len(options.Env) != 1 || !regexp.MustCompile(`^ODAC_TTY_SESSION=[a-f0-9]{32}$`).MatchString(options.Env[0]) {
		t.Errorf("Env = %v", options.Env)
	}
	if strings.Join(options.Cmd, " ") != strings.Join(DefaultShell, " ") {
		t.Errorf("Cmd = %v", options.Cmd)
	}

	// Tty must be repeated on attach or the daemon streams multiplexed
	// frames.
	f.tmu.Lock()
	attach := f.attachOptions[0]
	f.tmu.Unlock()
	if !attach.Tty {
		t.Error("Tty not repeated in exec attach")
	}
}

func TestTerminalOptionsPassThrough(t *testing.T) {
	f := newTermFakeAPI()
	term := openTerminal(t, f, TerminalOptions{
		Command: []string{"/bin/zsh"},
		User:    "app",
		Workdir: "/srv",
		Env:     []string{"FOO=bar"},
	})
	defer term.Close("closed")

	options := f.sessionOptions(t)
	if strings.Join(options.Cmd, ",") != "/bin/zsh" {
		t.Errorf("Cmd = %v", options.Cmd)
	}
	if options.User != "app" || options.WorkingDir != "/srv" {
		t.Errorf("User/WorkingDir = %s/%s", options.User, options.WorkingDir)
	}
	if len(options.Env) != 2 || !strings.HasPrefix(options.Env[0], "ODAC_TTY_SESSION=") || options.Env[1] != "FOO=bar" {
		t.Errorf("Env = %v", options.Env)
	}
}

func TestTerminalIO(t *testing.T) {
	f := newTermFakeAPI()
	var chunks []string
	var chunksMu sync.Mutex
	term := openTerminal(t, f, TerminalOptions{OnData: func(b []byte) {
		chunksMu.Lock()
		chunks = append(chunks, string(b))
		chunksMu.Unlock()
	}})

	// Output → onData.
	f.tmu.Lock()
	server := f.sessionServer
	f.tmu.Unlock()
	go server.Write([]byte("hello"))
	deadline := time.Now().Add(2 * time.Second)
	for time.Now().Before(deadline) {
		chunksMu.Lock()
		got := strings.Join(chunks, "")
		chunksMu.Unlock()
		if got == "hello" {
			break
		}
		time.Sleep(5 * time.Millisecond)
	}
	chunksMu.Lock()
	if strings.Join(chunks, "") != "hello" {
		t.Fatalf("chunks = %v", chunks)
	}
	chunksMu.Unlock()

	// Input → pty.
	read := make(chan string, 1)
	go func() {
		buf := make([]byte, 16)
		n, _ := server.Read(buf)
		read <- string(buf[:n])
	}()
	if !term.Write([]byte("ls\n")) {
		t.Fatal("write returned false")
	}
	select {
	case got := <-read:
		if got != "ls\n" {
			t.Fatalf("pty received %q", got)
		}
	case <-time.After(2 * time.Second):
		t.Fatal("pty never received input")
	}

	// Resize maps cols/rows onto Docker w/h.
	if !term.Resize(100, 30) {
		t.Fatal("resize returned false")
	}
	f.tmu.Lock()
	last := f.resizes[len(f.resizes)-1]
	f.tmu.Unlock()
	if last.Height != 30 || last.Width != 100 {
		t.Fatalf("resize = %+v", last)
	}

	// Nonsense dimensions clamp instead of reaching Docker raw.
	term.Resize(99999, 0)
	f.tmu.Lock()
	last = f.resizes[len(f.resizes)-1]
	f.tmu.Unlock()
	if last.Width != 1000 || last.Height != 30 {
		t.Fatalf("clamped resize = %+v", last)
	}

	// A shell that exited mid-resize is not an error worth raising.
	f.tmu.Lock()
	f.resizeErr = notFoundErr{"no such exec"}
	f.tmu.Unlock()
	if term.Resize(90, 20) {
		t.Fatal("resize survived a dead exec")
	}

	// No-ops once closed.
	term.Close("closed")
	if term.Write([]byte("x")) {
		t.Fatal("write after close")
	}
	if term.Resize(90, 20) {
		t.Fatal("resize after close")
	}
}

func TestTerminalCloseReaps(t *testing.T) {
	f := newTermFakeAPI()
	rec := &exitRecorder{}
	term := openTerminal(t, f, TerminalOptions{User: "app", OnExit: rec.record})

	term.Close("closed")

	scripts := f.scripts()
	if len(scripts) != 1 {
		t.Fatalf("reap scripts = %d", len(scripts))
	}
	script := scripts[0]
	if !strings.Contains(script, "kill -HUP") {
		t.Error("no SIGHUP in reap script")
	}
	// Single line: `done \n | sort` is a shell syntax error.
	if strings.Contains(script, "\n") {
		t.Error("reap script is multi-line")
	}
	// Targets only this session's tag.
	tag := strings.TrimPrefix(f.sessionOptions(t).Env[0], "ODAC_TTY_SESSION=")
	if !strings.Contains(script, "ODAC_TTY_SESSION="+tag) {
		t.Error("reap script does not carry the session tag")
	}
	// Children first (highest pid first), the shell last, with the grace
	// pause between.
	if !strings.Contains(script, "sort -rn") {
		t.Error("pids not sorted -rn")
	}
	if strings.Index(script, `kill -HUP "$p"`) > strings.Index(script, `kill -HUP "$shell"`) {
		t.Error("shell signalled before its children")
	}
	if !regexp.MustCompile(`kill -HUP "\$p".*sleep 1.*kill -HUP "\$shell"`).MatchString(script) {
		t.Error("no pause between children and shell")
	}

	// The reaper exec is untagged (never matches itself) and runs as the
	// session user (environ of another uid needs CAP_SYS_PTRACE).
	reap := f.reapOptions(t)
	if len(reap.Env) != 0 {
		t.Errorf("reap exec is tagged: %v", reap.Env)
	}
	if reap.User != "app" {
		t.Errorf("reap user = %q", reap.User)
	}

	// onExit fires exactly once with the exit code, even on repeated close.
	term.Close("closed")
	calls := rec.list()
	if len(calls) != 1 || calls[0].Reason != "closed" || calls[0].ExitCode != 0 {
		t.Fatalf("onExit calls = %+v", calls)
	}
	if !term.Closed() {
		t.Fatal("terminal not closed")
	}
}

func TestTerminalReapEscalatesToSIGKILL(t *testing.T) {
	f := newTermFakeAPI()
	// Running after HUP, still running after the grace wait, dead once
	// KILLed.
	f.runningSeq = []bool{true, true, false, false}
	term := openTerminal(t, f, TerminalOptions{})

	term.Close("closed")

	scripts := f.scripts()
	if len(scripts) != 2 {
		t.Fatalf("reap scripts = %d", len(scripts))
	}
	if !strings.Contains(scripts[0], "kill -HUP") || !strings.Contains(scripts[1], "kill -KILL") {
		t.Fatalf("escalation = %v", scripts)
	}
}

func TestTerminalShellExitClosesAsExited(t *testing.T) {
	f := newTermFakeAPI()
	rec := &exitRecorder{}
	openTerminal(t, f, TerminalOptions{OnExit: rec.record})

	f.tmu.Lock()
	f.sessionServer.Close() // the shell hangs up
	f.tmu.Unlock()

	deadline := time.Now().Add(2 * time.Second)
	for time.Now().Before(deadline) && len(rec.list()) == 0 {
		time.Sleep(5 * time.Millisecond)
	}
	calls := rec.list()
	if len(calls) != 1 || calls[0].Reason != "exited" || calls[0].ExitCode != 0 {
		t.Fatalf("onExit = %+v", calls)
	}
}

func TestTerminalSurvivesVanishedContainer(t *testing.T) {
	f := newTermFakeAPI()
	rec := &exitRecorder{}
	term := openTerminal(t, f, TerminalOptions{OnExit: rec.record})

	f.tmu.Lock()
	f.execCreateErr = notFoundErr{"No such container"}
	f.tmu.Unlock()

	term.Close("closed")
	if calls := rec.list(); len(calls) != 1 {
		t.Fatalf("onExit = %+v", calls)
	}
	if !term.Closed() {
		t.Fatal("terminal not closed")
	}
}

// drainPTY keeps reading the pty input side: net.Pipe writes are fully
// synchronous, so an undrained server would block term.Write (a real
// hijacked docker conn has kernel buffers).
func drainPTY(f *termFakeAPI) {
	f.tmu.Lock()
	server := f.sessionServer
	f.tmu.Unlock()
	go func() {
		buf := make([]byte, 1024)
		for {
			if _, err := server.Read(buf); err != nil {
				return
			}
		}
	}()
}

func TestTerminalIdleTimeout(t *testing.T) {
	f := newTermFakeAPI()
	rec := &exitRecorder{}
	term := openTerminal(t, f, TerminalOptions{
		OnExit:      rec.record,
		IdleTimeout: durPtr(120 * time.Millisecond),
	})
	drainPTY(f)

	// Input pushes the deadline back.
	time.Sleep(80 * time.Millisecond)
	term.Write([]byte("x"))
	time.Sleep(80 * time.Millisecond)
	if len(rec.list()) != 0 {
		t.Fatal("closed despite fresh input")
	}

	deadline := time.Now().Add(2 * time.Second)
	for time.Now().Before(deadline) && len(rec.list()) == 0 {
		time.Sleep(10 * time.Millisecond)
	}
	calls := rec.list()
	if len(calls) != 1 || calls[0].Reason != "idle" {
		t.Fatalf("onExit = %+v", calls)
	}
}

func TestTerminalOutputDoesNotResetIdle(t *testing.T) {
	f := newTermFakeAPI()
	rec := &exitRecorder{}
	openTerminal(t, f, TerminalOptions{
		OnExit:      rec.record,
		IdleTimeout: durPtr(100 * time.Millisecond),
	})

	// A chatty process must not keep an abandoned session alive.
	stop := make(chan struct{})
	go func() {
		for {
			select {
			case <-stop:
				return
			default:
				f.tmu.Lock()
				server := f.sessionServer
				f.tmu.Unlock()
				server.SetWriteDeadline(time.Now().Add(20 * time.Millisecond))
				server.Write([]byte("output"))
				time.Sleep(10 * time.Millisecond)
			}
		}
	}()
	defer close(stop)

	deadline := time.Now().Add(2 * time.Second)
	for time.Now().Before(deadline) && len(rec.list()) == 0 {
		time.Sleep(10 * time.Millisecond)
	}
	calls := rec.list()
	if len(calls) != 1 || calls[0].Reason != "idle" {
		t.Fatalf("onExit = %+v", calls)
	}
}

func TestTerminalLifetimeCap(t *testing.T) {
	f := newTermFakeAPI()
	rec := &exitRecorder{}
	term := openTerminal(t, f, TerminalOptions{
		OnExit:      rec.record,
		IdleTimeout: durPtr(100 * time.Millisecond),
		MaxLifetime: durPtr(250 * time.Millisecond),
	})
	drainPTY(f)

	// Steady input keeps resetting the idle timer; only the cap can end it.
	stop := make(chan struct{})
	go func() {
		for {
			select {
			case <-stop:
				return
			default:
				term.Write([]byte("x"))
				time.Sleep(50 * time.Millisecond)
			}
		}
	}()
	defer close(stop)

	deadline := time.Now().Add(3 * time.Second)
	for time.Now().Before(deadline) && len(rec.list()) == 0 {
		time.Sleep(10 * time.Millisecond)
	}
	calls := rec.list()
	if len(calls) != 1 || calls[0].Reason != "lifetime" {
		t.Fatalf("onExit = %+v", calls)
	}
}

func TestTerminalZeroTimeoutsDisable(t *testing.T) {
	f := newTermFakeAPI()
	rec := &exitRecorder{}
	term := openTerminal(t, f, TerminalOptions{OnExit: rec.record}) // helper zeroes both

	time.Sleep(150 * time.Millisecond)
	if len(rec.list()) != 0 {
		t.Fatal("timers fired despite 0 timeouts")
	}
	term.Close("closed")
}

// ---- CreateTerminalSession gates (Container.js side) ----

func runningInspect(privileged bool, user string) container.InspectResponse {
	return container.InspectResponse{
		ContainerJSONBase: &container.ContainerJSONBase{
			State:      &container.State{Running: true},
			HostConfig: &container.HostConfig{Privileged: privileged},
		},
		Config: &container.Config{User: user},
	}
}

func newTermClient(f *termFakeAPI, apps ...string) *Client {
	c := New(f, Options{})
	c.SetAppNames(func() []string { return apps })
	return c
}

func TestCreateTerminalSessionGates(t *testing.T) {
	f := newTermFakeAPI()
	f.inspects["web"] = runningInspect(false, "")
	c := newTermClient(f, "web")

	// Unknown app: never inspected, never exec'd.
	if _, err := c.CreateTerminalSession("odac", TerminalOptions{}); err == nil ||
		err.Error() != "Unknown app: odac" {
		t.Fatalf("unknown app err = %v", err)
	}

	// Not running (no inspect entry → 404).
	c2 := newTermClient(f, "ghost")
	if _, err := c2.CreateTerminalSession("ghost", TerminalOptions{}); err == nil ||
		err.Error() != "Container ghost is not running" {
		t.Fatalf("not-running err = %v", err)
	}

	// Privileged refusal without the opt-in.
	f.inspects["priv"] = runningInspect(true, "")
	c3 := newTermClient(f, "priv")
	if _, err := c3.CreateTerminalSession("priv", TerminalOptions{}); err == nil ||
		err.Error() != "Refusing a terminal in privileged container priv" {
		t.Fatalf("privileged err = %v", err)
	}
	term, err := c3.CreateTerminalSession("priv", TerminalOptions{AllowPrivileged: true, IdleTimeout: durPtr(0), MaxLifetime: durPtr(0)})
	if err != nil {
		t.Fatalf("privileged opt-in failed: %v", err)
	}
	term.Close("closed")
}

func TestCreateTerminalSessionUsesImageUser(t *testing.T) {
	f := newTermFakeAPI()
	f.inspects["web"] = runningInspect(false, "appuser")
	c := newTermClient(f, "web")

	term, err := c.CreateTerminalSession("web", TerminalOptions{IdleTimeout: durPtr(0), MaxLifetime: durPtr(0)})
	if err != nil {
		t.Fatal(err)
	}
	defer term.Close("closed")
	if f.sessionOptions(t).User != "appuser" {
		t.Errorf("exec user = %q", f.sessionOptions(t).User)
	}
}

func TestCreateTerminalSessionDockerDown(t *testing.T) {
	f := newTermFakeAPI()
	f.pingErr = notFoundErr{"down"}
	c := New(f, Options{})
	if _, err := c.CreateTerminalSession("web", TerminalOptions{}); err == nil ||
		err.Error() != "Docker is not available" {
		t.Fatalf("docker-down err = %v", err)
	}
}
